Publishing workflows for snippets

laymonage opened the pull request:
This PR implements publishing workflows for snippets.

<!--
Thanks for contributing to Wagtail! 🎉

Before submitting, please review the [contributor guidelines](https://docs.wagtail.org/en/latest/contributing/index.html).
-->

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [x] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

To enable workflows for snippets, add `WorkflowMixin` to your snippet model. It also requires `LockableMixin`, `DraftStateMixin`, and `RevisionMixin` to be enabled. Once enabled, open the Workflow edit view through Settings > Workflows, a new section should show up where you can assign the workflow to snippet models that have `WorkflowMixin` enabled.

For models that have a workflow assigned, the create and edit view will show the workflow information and action menu items. The rest of the feature works exactly as it works for pages.

The dashboard will show any snippets (alongside pages) that are in review or pending your review. The Workflow and Workflow Task reports will also show the reports for snippets.

I made a branch on my fork of bakerydemo to test things with the bread ingredient and person snippets: https://github.com/laymonage/bakerydemo/tree/snippets-workflow

Some commits may worth backporting, e.g.:
- 6635254c3ef51cea922bb66a348c3ced38456c14
- f4621b4f02c21fcd9b49072838cc44cd516ab86e
- 3bcb348f76cf35243851eb2d97f1482924b652cf
- 57bbb59375175e9cd66abbd13f859eaeb74b1777
- 38c112ceaf9843895440066e8ce1c8d680002d91



[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

laymonage commented on wagtail/snippets/templates/wagtailsnippets/snippets/edit.html:
```diff
@@ -62,6 +62,38 @@
             });
 
             LockUnlockAction('{{ csrf_token|escapejs }}', '{{ action_url }}');
+            ActivateWorkflowActionsForEditView('[data-edit-form]');
+
+            {% if publishing_will_cancel_workflow %}
+                /* Make user confirm before publishing the snippet if it will cancel an ongoing workflow */
```
Really wish I didn't have to add more to the inline scripts, but I'm just copying this from pages and I think we can refactor them both together later, possibly with Stimulus

nmorduch reviewed (commented):
@jacobtoppm and I started review on this commit by commit.

@laymonage there's no need to respond to any of the comments yet, since we had to stop partway through the review and the comments are all questions that may be resolved after reviewing the rest.

jacobtoppm commented on wagtail/admin/forms/workflows.py:
```diff
@@ -149,6 +152,98 @@ def clean(self):
+            WorkflowContentType(workflow=self.workflow, content_type=ct)
+            for ct in content_types
+        ]
+        WorkflowContentType.objects.bulk_create(objects, ignore_conflicts=True)
```
Why ignore conflicts?

jacobtoppm replied:
Thought it might be something like that - makes sense, thanks.

jacobtoppm commented on wagtail/admin/templates/wagtailadmin/workflows/includes/workflow_content_types_checkbox.html:
```diff
@@ -0,0 +1,14 @@
+{% load wagtailadmin_tags %}
+
+{% for id, errors in errors_by_id.items %}
+    {% if id == widget.value %}
+        <div class="w-field__errors" data-field-errors>
+            {% icon name="warning" class_name="w-field__errors-icon" %}
+            <p class="error-message">
+                {% for error in errors %}{{ error.message }} {% endfor %}
+            </p>
+        </div>
+    {% endif %}
+{% endfor %}
```
Yeah, I've run across wanting a tag or filter like this before too. I wouldn't be averse to adding it if we come across another use in Wagtail

jacobtoppm commented on wagtail/admin/forms/workflows.py:
```diff
@@ -149,6 +152,98 @@ def clean(self):
+            *get_workflow_enabled_models()
+        ).values()
+        self.fields["content_types"].queryset = ContentType.objects.filter(
+            pk__in=[ct.pk for ct in workflow_cts]
```
I don't love the double query here, though it's probably not worth nitpicking over a single additional query! But you could get it in one by just querying directly on app label and model name for each (combining Q objects to represent the ANDs and ORs correctly)

jacobtoppm commented on wagtail/snippets/action_menu.py:
```diff
@@ -77,6 +78,110 @@ def is_shown(self, context):
+            return context["model"].get_default_workflow() is not None
+
+        return (
+            not context.get("locked_for_user")
```
Could we be showing this on the revisions_revert view here, as the corresponding page logic suggests?

laymonage replied:
Correct, we shouldn't show it, I missed this. Thanks!

jacobtoppm commented on wagtail/snippets/action_menu.py:
```diff
@@ -77,6 +78,121 @@ def is_shown(self, context):
+    label = _("Submit for moderation")
+    icon_name = "resubmit"
+
+    def is_shown(self, context):
```
Yes, I think these share a huge amount of logic - it would be really good to reduce that duplication. I do wonder if you could could just refactor out the few page-specific checks into separate methods quite easily here?

laymonage replied:
The classes are currently not inherited nor have any shared templates whatsoever with the ones from pages, so even if refactored into separate methods, it still requires a fair amount of other refactoring to share the logic.

jacobtoppm commented on wagtail/admin/views/generic/mixins.py:
```diff
@@ -221,24 +225,79 @@ def setup(self, request, *args, **kwargs):
+        if not self.workflow_enabled or not self.object:
+            return
+
+        self.workflow_state = (
```
Do we always need these (this and below), or could they be cached properties only calculated when needed?

laymonage replied:
I think we can calculate the `workflow_tasks` later, but the rest are used pretty early in the flow to determine the available actions.

jacobtoppm commented on wagtail/snippets/action_menu.py:
```diff
@@ -298,6 +299,32 @@ def __init__(self, request, **kwargs):
             if menu_item.is_shown(self.context)
         ]
 
+        if instance and isinstance(instance, WorkflowMixin):
```
Isn't this pretty much identical to the `Page` logic? Could it be shared?

laymonage replied:
Yes, but at this point I think it's too risky to unify them together. We have an existing issue to redesign this (#9030, #9869), which I think would be an excellent time to unify the two.

jacobtoppm reviewed (commented):
WIP review so far

jacobtoppm commented on wagtail/snippets/templates/wagtailsnippets/snippets/action_menu/workflow_menu_item.html:
```diff
@@ -0,0 +1,4 @@
+{% load wagtailadmin_tags %}
```
Why can't this be shared?

laymonage replied:
As noted in the comment above, we're currently not sharing any of the action menu items, and I think it's best to unify them both together later.

jacobtoppm commented on wagtail/snippets/templates/wagtailsnippets/snippets/edit.html:
```diff
@@ -63,6 +63,37 @@
             LockUnlockAction('{{ csrf_token|escapejs }}', '{{ action_url }}');
             ActivateWorkflowActionsForEditView('[data-edit-form]');
+
+            {% if publishing_will_cancel_workflow %}
```
Again, could we share this rather than copy it, if the logic is identical aside from the view?

jacobtoppm commented on wagtail/admin/views/home.py:
```diff
@@ -118,10 +119,23 @@ def get_context_data(self, parent_context):
+                Page.objects.filter(
+                    owner=request.user,
+                    id=Cast(OuterRef("object_id"), output_field=IntegerField()),
+                ).values_list(Cast("id", output_field=CharField()), flat=True)
```
I'm a bit confused why this returns an id - isn't it essentially a boolean check you're doing here?

laymonage replied:
I'm not sure I follow -- is there a way to reduce this to a boolean check?

I needed to do:
- from the active workflow states,
- filter where the generic foreign key point to pages that are owned by the user

jacobtoppm replied:
Ah, sorry I wasn't clear enough here - your change is indeed exactly what I meant, I was just querying why we were doing a subquery to return a list of 1/0 ids per row and then checking if the main query's id was in that, vs an `Exists` or similar.

laymonage commented on wagtail/admin/views/reports/workflows.py:
```diff
@@ -147,19 +147,19 @@ class WorkflowTasksView(ReportView):
     export_headings = {
-        "workflow_state.page.id": _("Page ID"),
-        "workflow_state.page.content_type.model_class._meta.verbose_name.title": _(
+        "workflow_state.content_object.id": _("Page ID"),
```
Thanks, I've fixed the issues with the export stuff.

laymonage commented on wagtail/models/__init__.py:
```diff
@@ -4055,7 +4057,7 @@ def cancel(self, user=None):
 
         log(
-            instance=self.page.specific,
+            instance=self.content_object.specific,
```
This seems to be from one of the older commits, I've changed this in the recent commits to have a `isinstance(self.content_object, Page)` check before using `.specific`.

jacobtoppm replied:
Thanks!

laymonage commented on wagtail/models/__init__.py:
```diff
@@ -783,6 +783,127 @@ def get_preview_template(self, request, mode_name):
 
+class WorkflowMixin:
+    @classmethod
+    def get_default_workflow(cls):
```
Not sure... It won't be called anywhere on a Page by wagtail, but I don't see why we should stop developers (if they so wish) to make use of this method to have a default workflow based on the specific Page model class rather than the `WorkflowPage` link. They'll probably need to change some other stuff as well, but I'd assume they'll know what they're doing anyway.

laymonage commented on wagtail/models/__init__.py:
```diff
@@ -980,6 +980,11 @@ def get_lock(self):
         if self.locked:
             return BasicLock(self)
 
+        if isinstance(self, WorkflowMixin) and self.current_workflow_task:
```
I suppose we could define `get_lock()` in `WorkflowMixin` as well, and use `super().get_lock()`. However, I was thinking of the idea that the `LockableMixin` class being the one who knows all about the default locks we have, and since we already have a `DraftStateMixin` check inside here (even though both are completely independent), might as well have one for `WorkflowMixin` here...

Though we could also argue that we do state that the `WorkflowMixin` requires `LockableMixin`, so we could assume that `LockableMixin` is applied and we could call `super().get_lock()` from `WorkflowMixin`. Maybe that's better. I'll have another look tomorrow

jacobtoppm replied:
If we've gone that way for `DraftStateMixin`, maybe that makes sense then. The pattern of mixins checking which other mixins are applied feels slightly awkward in places - here I think the method + `super()` in `WorkflowMixin` would work for the logic, did it not work neatly for `DraftStateMixin` then?

laymonage replied:
With `DraftStateMixin`, locking is completely optional (and was introduced later when we added `LockableMixin` support for snippets), thus I felt like it shouldn't need to know anything about a `get_lock()` method. I initially thought it would be more readable if the lock is always returned from `LockableMixin`, so we know what the order of precedence is.

However, now that we say `WorkflowMixin` requires `LockableMixin`, it does feel a bit weird for `LockableMixin` to know about `WorkflowMixin`, so I guess using `super()` here works better.

Unfortunately, if we'd like to do the same for `DraftStateMixin`, we'll have to add another check to ensure that the MRO for `DraftStateMixin` comes first before `LockableMixin` (or, we add a `super().get_lock()` in `LockableMixin` itself, which seems a bit silly). We didn't add any explicit checks to ensure this when we introduced `LockableMixin` for snippets because both mixins can be used independently. If we want to do this, now would be the best time before the RC is out... But I feel like it's out of the scope of this PR?

In the interest of time, I'd say we make `WorkflowMixin` use `super().get_lock()`, but leave the check for `DraftStateMixin`. This way, we could make `LockableMixin` not rely on `WorkflowMixin`, but keep the "canonical" order of the mixins exactly as the order they were introduced: `WorkflowMixin`, `LockableMixin`, `DraftStateMixin`, `RevisionMixin`.

jacobtoppm replied:
Yep, I agree with that. That said - is there anything intrinsic to `LockableMixin` required by `WorkflowMixin` having a `get_lock` method? It doesn't use any of the fields on that mixin, for example. So in theory it could be independent, but would rely on views/other areas checking for a `get_lock` method rather than `LockableMixin`,

For now, yeah, let's go with `WorkflowMixin` defines its own `get_lock` method and calls `super()` - I think that makes most sense

laymonage replied:
> Yep, I agree with that. That said - is there anything intrinsic to `LockableMixin` required by `WorkflowMixin` having a `get_lock` method? It doesn't use any of the fields on that mixin, for example. So in theory it could be independent, but would rely on views/other areas checking for a `get_lock` method rather than `LockableMixin`,

Nope, you're right there isn't. It's just that I'm not sure if that's something we want to support officially. It's much easier to maintain if we say that you need `LockableMixin` when using `WorkflowMixin` from the get-go instead of having to guard against all possible combinations. Though I'll see if I can make the workflow code safer without `LockableMixin` applied, and see if this is something we can support.

laymonage commented on wagtail/locks.py:
```diff
@@ -151,22 +149,46 @@ def get_message(self, user):
+                if self.is_page:
+                    workflow_info = _("This page is currently awaiting moderation.")
+                else:
+                    workflow_info = capfirst(
```
> Is it in case the translation has model name first and the model name isn't capitalized?

Yep

> Does capfirst know that some languages don't have capitalization?

It uses Python's `upper()` so it should handle it gracefully. If the language doesn't have capitalisation then it doesn't matter, no?

jacobtoppm replied:
Yep, this was just a check. As mentioned, these were more notes to double check later.

laymonage commented on wagtail/locks.py:
```diff
@@ -151,22 +149,46 @@ def get_message(self, user):
+                    )
+                    # Make sure message is correctly capitalised even if it
+                    # starts with model_name.
+                    workflow_info = mark_safe(capfirst(workflow_info))
```
Same reason as above, and as written in the comment above the code

jacobtoppm replied:
Yeah, saw the comment, was just confirming this was in case the translation rearranged the message, since the raw message doesn't start with model name

jacobtoppm commented on wagtail/admin/views/generic/history.py:
```diff
@@ -0,0 +1,173 @@
+    def dispatch(self, request, *args, **kwargs):
+        # Do these here instead of setup() so that subclasses can do permission
+        # checking using self.object before these are called
+        self.workflow_state = self.get_workflow_state()
```
Why are we retrieving these here rather than where we need them?

jacobtoppm commented on wagtail/admin/templates/wagtailadmin/notifications/task_state_approved.html:
```diff
@@ -3,6 +3,26 @@
+            {% blocktrans trimmed with title=title task=task.name model_name=model_name asvar message %}
+                The {{ model_name }} "{{ title }}" has been approved in moderation stage "{{ task }}".
+            {% endblocktrans %}
+            {{ message|capfirst|safe }}
```
Does the message need to be marked as safe here? I can't see any html called for.

laymonage replied:
You're right for the HTML files, but for the `.txt` files we need them as otherwise characters like `"` get escaped as `&quot;`, which will show up as-is in the email.

jacobtoppm commented on wagtail/admin/templates/wagtailadmin/notifications/task_state_approved.html:
```diff
@@ -3,6 +3,26 @@
+                {% blocktrans trimmed with model_name=model_name asvar edit_message %}
+                    You can edit the {{ model_name }} here:
+                {% endblocktrans %}
+                {{ edit_message|capfirst|safe }}
```
Same here.

jacobtoppm commented on wagtail/admin/templates/wagtailadmin/notifications/task_state_approved_subject.txt:
```diff
@@ -1,3 +1,11 @@
+    {% blocktrans trimmed with title=title|safe task=task.name|safe model_name=model_name|safe asvar message %}
+        The {{ model_name }} "{{ title }}" has been approved in "{{ task }}".
+    {% endblocktrans %}
+    {{ message|capfirst|safe }}
```
And here, but I'll stop noting all the templates now - just a generic question about the use of `safe` for all.

jacobtoppm commented on wagtail/admin/views/generic/mixins.py:
```diff
@@ -256,18 +256,36 @@ def setup_workflow(self):
-        # if they are the user who locked the object
-        if permission == "unlock" and self.object.locked_by_id == self.request.user.pk:
+        # Workflow lock/unlock methods take precedence before the base
+        # "lock" and "unlock" permissions -- see PagePermissionTester for reference
```
Not a big deal for this PR necessarily, but I wonder if moving towards a view-level permission checking here, vs a tester class in Page is causing the permissions system to diverge further. I think it would be nice to consolidate how we do this. (eg a class supplied by the permission policy?

jacobtoppm replied:
(more a general point for the future - I really feel like we should consolidate permissions checks)

laymonage replied:
I agree, I had the same concern and yep I think this needs its own dedicated time for API design and implementation.

jacobtoppm commented on wagtail/admin/forms/workflows.py:
```diff
@@ -149,6 +152,98 @@ def clean(self):
+        def label_from_instance(self, obj):
+            return capfirst(obj.name)
+
+    class CheckboxSelectMultiple(forms.CheckboxSelectMultiple):
+        """Custom CheckboxSelectMultiple widget that renders errors for each content type ID"""
+
+        option_template_name = (
+            "wagtailadmin/workflows/includes/workflow_content_types_checkbox.html"
+        )
+
+        def get_errors_by_id(self, errors):
+            errors_by_id = {}
+            for error in errors.as_data():
+                ct_id = error.params and error.params.get("content_type_id")
+                errors_by_id.setdefault(ct_id, []).append(error)
+            return errors_by_id
+
+        def render_with_errors(
+            self, name, value, attrs=None, renderer=None, errors=None
+        ):
+            context = {
+                **self.get_context(name, value, attrs),
+                "errors_by_id": self.get_errors_by_id(errors),
+            }
+            return self._render(self.template_name, context, renderer)
```
Makes sense!

jacobtoppm commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -156,6 +161,57 @@ def page_permissions(context, page):
+
+
+@register.simple_tag(takes_context=True)
+def admin_edit_url(context, obj, user=None):
```
I think some of it may just be the fact that `AdminURLFinder` was added more recently (in Wagtail's terms) than many other features, to make audit logging easier I believe

jacobtoppm commented on wagtail/admin/views/workflows.py:
```diff
@@ -31,9 +33,11 @@
     WorkflowState,
 )
 from wagtail.permissions import task_permission_policy, workflow_permission_policy
+from wagtail.snippets.models import get_workflow_enabled_models
```
I think it's okay for now, at least until we unify snippets and pages a bit more

jacobtoppm commented on wagtail/locks.py:
```diff
@@ -151,22 +149,46 @@ def get_message(self, user):
             if len(self.object.current_workflow_state.all_tasks_with_status()) == 1:
                 # If only one task in workflow, show simple message
-                workflow_info = _("This page is currently awaiting moderation.")
+                if self.is_page:
```
That makes sense.

laymonage commented:
@jacobtoppm I've pushed new commits that addressed most comments, but I'll write a summary tomorrow to help you re-review this and I'll see if I can write more tests to cover the gotchas you highlighted. Thanks!

jacobtoppm commented on wagtail/models/__init__.py:
```diff
@@ -864,6 +864,60 @@ def get_lock(self):
+        try:
+            # WorkflowMixin calls super().get_lock() in the override,
+            # so if applied, it must be applied before LockableMixin.
+            if mro.index(LockableMixin) < mro.index(WorkflowMixin):
```
Hmm - I think our usages of "applied before" and "after" are at odds here! I'd think "applied before" meant "applied closer to the base class", ie to the right, and "applied after" would mean it was closer to being the final layer in the inheritance, so closer to the left. That said, our other errors go with this wording, so maybe it's just me who reads it that way! Definitely not a blocker, but maybe something for us to look at rewording how we write inheritance errors if anyone else turns out to find this confusing.

laymonage replied:
Yeah this is tricky, I initially had the same thought as yours when I wrote the first check for `DraftStateMixin`. The thing is, you can say it can be confusing if we go that way as well: "you say it needs to be *applied before*, but why do you actually need to *write it after*?"

I decided following the MRO for the "before/after" makes more sense here, because it's consistent:
- If you write `WorkflowMixin, LockableMixin` let's state this as applying `WorkflowMixin` *before* `LockableMixin`
- Methods from `WorkflowMixin` will be called *before* those from `LockableMixin`
- When you provide the hint to "put `WorkflowMixin` before `LockableMixin`", it's exactly what it reads: `WorkflowMixin, LockableMixin`.

jacobtoppm replied:
Yeah, I suspect it'll be obvious in context anyway - if it's complaining about the model inheritance being one way round, it's probably natural to try the other way

jacobtoppm commented on wagtail/models/__init__.py:
```diff
@@ -780,6 +783,141 @@ def get_preview_template(self, request, mode_name):
+        """
+        return WorkflowState.objects.for_instance(self)
+
+    @property
```
@laymonage is this something you're still intending to add? Or in a future PR?

laymonage replied:
I think I'll do this in a future PR, with some query analysis to back up the rationale and adjustments to the tests to make sure everything works properly.

laymonage commented:
Hey @jacobtoppm, thanks for reviewing again! Sorry I didn't let you know before you did, but in general:

- Most changes are made in new commits.
- I amended the commits where `WorkflowMixin` is first introduced to make it written *below* `LockableMixin`, so it makes sense to use `super().get_lock()` and the way the inheritance is enforced (`WorkflowMixin, LockableMixin`). I didn't create a new commit for this because doing that would erase the `LockableMixin` history (due to how git works out the diff)
- https://github.com/wagtail/wagtail/pull/9817/commits/7f798d687b2a15501ac642336943c3f9b5017990 was a bit different when you re-reviewed, I changed this to use cached properties
- https://github.com/wagtail/wagtail/pull/9817/commits/3acb25057bdab47a8d879001fd7914211b49e9c4 same as above

but yeah, I addressed all of the second review comments in new commits after https://github.com/wagtail/wagtail/pull/9817/commits/5ab0a8502f8b83ecbe18efbb6b9bd50ac4de6c2e

(I rebased from latest `main` though, as there was a conflict with one of the templates)

laymonage reviewed (commented):
Hey @jacobtoppm, I've addressed the comments as per our discussion, here's the highlights:

- f5902d8 For the system checks, I only check for `DraftStateMixin` and `RevisionMixin`, and not `LockableMixin`, as `WorkflowMixin` doesn't technically need it.
- a6880f3 Refactored the `get_lock()` method to handle the scheduled publishing lock from `DraftStateMixin`. I updated the docs to instruct that `LockableMixin` should be put after the other mixins but before the `RevisionMixin`. Technically, nothing should break if the developer put `LockableMixin` before the other mixins, but that means the workflow and scheduled publishing lock won't take effect. However, I did not add another system check for this, as I don't like the idea of making `LockableMixin` having to know all the other mixins in the system checks. I think the docs and existing check `LockableMixin < RevisionMixin` already suffices. 
- 3a4a598 Added tests for snippet workflows without `LockableMixin` applied. To avoid bloating the tests, I only run a subset of the tests where I think would be important, such as testing the approve/reject flows, but not the Workflow settings tests.
- 43e6a17 Added tests for making sure the action menu items are shown/hidden properly on both edit and revisions revert view.
- 1f9aad2 Added tests for the export feature of the workflow report views, for both pages and snippets (we didn't test the export before).

jacobtoppm reviewed (approved):
All looks good, thanks for your work on this! There is some followup work needed:
- A PR to cache more workflow queries
- More testing of how this feature impacts performance/query counts, and any fixes to address this
- Hopefully: refactoring action menu items in the future to duplicate less code
- In general: consideration of "how we do permissions" in the future, so we have one central approach, and refactoring to match

jacobtoppm commented on docs/releases/4.2.md:
```diff
@@ -245,3 +251,18 @@ If you are using the front-end cache invalidator module (`wagtail.contrib.fronte
+
+To accommodate workflows support for snippets, the `page` argument in {meth}`Workflow.start() <wagtail.models.Workflow.start>` has been renamed to `obj`.
+
+In addition, some methods on the base {class}`~wagtail.models.Task` model have been changed.
```
I think this is the most important bit, as we expect customisations here, whereas we don't elsewhere in workflow, so I think worth explicitly mentioning that this is relevant for custom task models. Also, I think you'll need to update https://github.com/laymonage/wagtail/blob/snippets-workflow/docs/extending/custom_tasks.md to match, and perhaps link it here.

jacobtoppm commented on docs/extending/custom_tasks.md:
````diff
@@ -124,34 +124,34 @@ class UserApprovalTask(Task):
 ```
 
-`Task.page_locked_for_user(page, user)`:
+`Task.locked_for_user(obj, user)`:
````
Worth adding a note about which version this changed in?

jacobtoppm reviewed (approved):
One tiny version change that would be nice to note, but otherwise newest changes look good!
